Add support for "irods" auth scheme, password hashing, and session tokens (main)

alanking opened the pull request:
Addresses irods/irods#2899
In service of irods/irods#8697

Companion PR: https://github.com/irods/irods/pull/8732

korydraughn commented on src/iadmin.cpp:
```diff
@@ -2056,12 +2076,23 @@ usage( char *subOpt ) {
+        "By default, the password is transferred in a scrambled form to be more secure.",
+        "Use the no-scramble option after the newValue parameter when setting the password",
+        "to indicate that the password should not be scrambled before being sent to the",
+        "server. This option is required when connected to servers running in FIPS mode.",
```
Is it "FIPS mode" or a "FIPS-compliant mode"?

Should we mention FIPS at all?
Do we need to come up with a name for FIPS related options?

Hmm...

alanking replied:
Sentence removed.

korydraughn commented on src/iadmin.cpp:
```diff
@@ -2585,6 +2615,24 @@ usage( char *subOpt ) {
+        "iadmin delete_session_tokens all (removes all session tokens)",
+        "iadmin delete_session_tokens expired (removes all expired session tokens)",
+        "iadmin delete_session_tokens all alice (removes all session tokens for user alice)",
+        "iadmin delete_session_tokens expired alice (removes all expired session tokens for user alice)",
```
I'm wondering if the descriptions should be on their own line, to guard against readers misinterpreting how to use these.

Consider looking at how various linux commands present examples.

alanking replied:
How does this look?
```suggestion
        "To remove all session tokens:",
        " $ iadmin delete_session_tokens all",
        "To remove all expired session tokens:",
        " $ iadmin delete_session_tokens expired",
        "To remove all session tokens for user alice:"
        " $ iadmin delete_session_tokens all alice",
        "To remove all expired session tokens for user alice:",
        " $ iadmin delete_session_tokens expired alice",
```

korydraughn replied:
That works.

korydraughn commented on src/ipasswd.cpp:
```diff
@@ -283,15 +310,23 @@ main( int argc, char **argv ) {
-    printf( "Usage: %s [-hvVl]\n", prog );
+    printf("Usage: %s [-ehfvVl]\n", prog);
+    printf("Changes your iRODS password.\n");
+    printf("Prompts for your old and new passwords.\n");
```
I think this line can be removed.

alanking commented on src/iadmin.cpp:
```diff
@@ -2056,12 +2076,23 @@ usage( char *subOpt ) {
+        "Use the no-scramble option after the newValue parameter when setting the password",
+        "to indicate that the password should not be scrambled before being sent to the",
+        "server. This option is required when connected to servers running in FIPS mode.",
+        "NOTE: The no-scramble option causes the password to be sent in plaintext.",
```
Done

alanking commented on src/iadmin.cpp:
```diff
@@ -2056,12 +2076,23 @@ usage( char *subOpt ) {
+        "NOTE: The no-scramble option causes the password to be sent in plaintext.",
+        "Please ensure TLS is in use when using this option. Using no-scramble with any",
+        "field other than password has no effect. Using no-scramble with servers before",
+        "iRODS 5.1.0 will result in users not being able to authenticate with that password.",
```
You would need to set the user's password again in order to get out of it. It is equivalent to setting your password to something that you don't know or forgot. What ends up happening is that the password gets descrambled in GeneralAdmin even though it's not scrambled, so it ends up being some nonsense string. Then that nonsense string gets re-scrambled before it's put into the catalog, and becomes the password.

korydraughn replied:
That's a possibility, but I think we have to say something about it somewhere unless ... we detect the server version and have the icommand reject the request if no-scramble is passed?

korydraughn replied:
Okay. Let's do that. That keeps people from entering that state via the icommands.

alanking commented on src/iexit.cpp:
```diff
@@ -72,27 +80,63 @@ main( int argc, char **argv ) {
+        if (remove_session_token_file) {
+            status = irods::authentication::remove_session_token_file();
+            if (myRodsArgs.verbose == True) {
+                std::printf("session token file unlink status: [%d]\n", status);
```
Done

alanking commented on src/igroupadmin.cpp:
```diff
@@ -461,47 +469,50 @@ printMsgs( char *msgs[] ) {
+                    " ",
+                    "A blank execute line invokes the interactive mode, where it",
+                    "prompts and executes commands until 'quit' or 'q' is entered.",
+                    "Single or double quotes can be used to enter items with blanks.",
```
Uhhh... looks like it:
```bash
$ igroupadmin -h | grep Version
iRODS Version 5.0.90                igroupadmin
$ igroupadmin
groupadmin>lg
public
groupadmin>q
```

alanking replied:
Created https://github.com/irods/irods/issues/8738 and https://github.com/irods/irods/issues/8739.

alanking commented on src/igroupadmin.cpp:
```diff
@@ -461,47 +469,50 @@ printMsgs( char *msgs[] ) {
+                          "Zone is the zone which the user is from (defaults to the local zone)",
+                          "no-scramble specifies that the password should not be scrambled before being sent",
+                          "to the server. This option is required when connected to servers running in FIPS",
+                          "mode. NOTE: The no-scramble option causes the password to be sent in plaintext.",
```
Done

alanking commented on src/igroupadmin.cpp:
```diff
@@ -461,47 +469,50 @@ printMsgs( char *msgs[] ) {
+                          "Zone is the zone which the user is from (defaults to the local zone)",
+                          "no-scramble specifies that the password should not be scrambled before being sent",
+                          "to the server. This option is required when connected to servers running in FIPS",
+                          "mode. NOTE: The no-scramble option causes the password to be sent in plaintext.",
```
I think we should just remove the mention of FIPS mode.

alanking commented on src/ipasswd.cpp:
```diff
@@ -283,15 +310,23 @@ main( int argc, char **argv ) {
-    printf( "password in a scrambled form to be used automatically by the icommands.\n" );
-    printf( "Prompts for your old and new passwords.\n" );
-    printf( "Usage: %s [-hvVl]\n", prog );
+    printf("Usage: %s [-ehfvVl]\n", prog);
```
Done. Also added `[--no-scramble]` to the list of options.

alanking commented on src/ipasswd.cpp:
```diff
@@ -310,9 +310,9 @@ main( int argc, char **argv ) {
 void usage( char *prog ) {
-    printf("Usage: %s [-ehfvVl]\n", prog);
+    printf("Usage: %s [-ehfvVl] [--no-scramble]\n\n", prog);
     printf("Changes your iRODS password.\n");
```
Done

korydraughn commented on src/iadmin.cpp:
```diff
@@ -1228,6 +1228,15 @@ doCommand( char *cmdToken[], rodsArguments_t* _rodsArgs = 0 ) {
+                const auto version = irods::to_version(Conn->svrVersion->relVersion);
+                if (version && version.value() < irods::version{5, 0, 90}) {
+                    fmt::print("no-scramble option is not supported before iRODS 5.1.0.\n");
+                    return 0;
```
Is there a way to indicate an error here?
Maybe there's an example of that in this function?

korydraughn replied:
Na, it's a known problem and can be handled in a separate PR.

alanking commented on src/igroupadmin.cpp:
```diff
@@ -296,6 +296,15 @@ doCommand( char *cmdToken[] ) {
+                if (version && version.value() < irods::version{5, 0, 90}) {
+                    fmt::print("no-scramble option is not supported before iRODS 5.1.0.\n");
+                    return 0;
+                }
```
Actually, I think this raises a valid point. Just seeing "no-scramble option is not supported before iRODS 5.1.0." after running the iCommand does not indicate to me whether or not it actually failed - it reads like a warning. So I'm going to put some verbiage around the request failing on each of these.

korydraughn reviewed (commented):
Squash to taste.

alanking commented:
Squashed. I added an extra commit at the end because some formatting changes happened during the rebase that I do not think are worth the time to resolve. The extra commit ensures that there is no difference before and after the rebase, so, no change has occurred since the last review comments.

alanking commented:
#'d, mergin
